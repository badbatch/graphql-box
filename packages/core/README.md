# @graphql-box/core

The GraphQL Box core module.

[![npm version](https://badge.fury.io/js/%40graphql-box%2Fcore.svg)](https://badge.fury.io/js/%40graphql-box%2Fcore)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](LICENSE)

## Installation

```bash
yarn add @graphql-box/core
```

## Usage

Read the GraphQL Box [usage guide](../../README.md#usage).

## Documentation

Read the module API [documentation](docs/README.md).
