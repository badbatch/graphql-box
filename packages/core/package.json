{
  "name": "@graphql-box/core",
  "version": "4.1.6",
  "description": "The GraphQL Box core module.",
  "author": "Dylan Aubrey <dylanaubrey@gmail.com>",
  "license": "MIT",
  "homepage": "https://github.com/badbatch/graphql-box",
  "bugs": {
    "url": "https://github.com/badbatch/graphql-box/issues"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/badbatch/graphql-box",
    "directory": "packages/core"
  },
  "publishConfig": {
    "access": "public"
  },
  "main": "lib/main/index.js",
  "module": "lib/module/index.mjs",
  "browser": "lib/browser/index.js",
  "types": "lib/types/index.d.ts",
  "keywords": [
    "graphql-box",
    "graphql",
    "client",
    "server",
    "isomorphic"
  ],
  "scripts": {
    "compile": "concurrently npm:compile:*",
    "compile-watch": "concurrently npm:compile-watch:*",
    "compile-watch:main": "yarn run compile:main --watch",
    "compile-watch:types": "yarn run compile:types --watch",
    "compile:browser": "cross-env BABEL_ENV=browser rollup -c ../../rollup.config.js",
    "compile:main": "cross-env BABEL_ENV=main babel ./src --out-dir ./lib/main --extensions \".ts\" --source-maps --config-file ../../babel.config.js",
    "compile:module": "cross-env BABEL_ENV=module babel ./src --out-dir ./lib/module --extensions \".ts\" --out-file-extension \".mjs\" --source-maps --config-file ../../babel.config.js",
    "compile:types": "tsc --build",
    "watch": "cross-env BABEL_ENV=browser rollup -c ../../rollup.config.js -w"
  },
  "dependencies": {
    "serialize-error": "8.1.0"
  },
  "peerDependencies": {
    "@babel/runtime": "< 8",
    "core-js": "< 4",
    "lodash": "< 5"
  }
}