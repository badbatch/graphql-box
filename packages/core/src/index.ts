import * as coreDefs from "./defs";
import Core from "./main";

export {
  Core as default,
  coreDefs,
};
