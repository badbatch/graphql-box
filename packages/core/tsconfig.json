{
  "extends": "../tsconfig.settings.json",
  "compilerOptions": {
    "outDir": "lib/types",
    "rootDir": "src"
  },
  "include": [
    "src"
  ]
}