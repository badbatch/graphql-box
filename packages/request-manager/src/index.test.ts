import { MaybeRawResponseData } from "@handl/core";
import {
  getRequestData,
  githubParsedQueries,
  githubQueryResponses,
} from "@handl/test-utils";
import fetchMock from "fetch-mock";
import { RequestManager } from ".";

const URL = "https://api.github.com/graphql";

describe("@handl/request-manager", () => {
  let requestManager: RequestManager;

  describe("when batching is set to false", () => {
    let response: MaybeRawResponseData;

    beforeAll(async () => {
      requestManager = await RequestManager.init({
        url: URL,
      });

      const body = { data: githubQueryResponses.singleType.data };
      const headers = { "cache-control": "public, max-age=5" };
      fetchMock.post("*", { body, headers });
      response = await requestManager.fetch(getRequestData(githubParsedQueries.singleType));
    });

    afterAll(() => {
      fetchMock.restore();
    });

    it("then the request manager should send the correct request", () => {
      expect(fetchMock.lastCall()).toMatchSnapshot();
    });

    it("then the request manager should return the correct response data", () => {
      expect(response).toMatchSnapshot();
    });
  });

  describe("when batching is set to true", () => {
    describe("when a single request is made within a batch interval", () => {
      let response: MaybeRawResponseData;

      beforeAll(async () => {
        jest.useFakeTimers();

        requestManager = await RequestManager.init({
          batch: true,
          fetchTimeout: 10000,
          url: URL,
        });

        const requestData = getRequestData(githubParsedQueries.singleType);

        const body = {
          batch: {
            [requestData.hash]: { data: githubQueryResponses.singleType.data },
          },
        };

        const headers = { "cache-control": "public, max-age=5" };
        fetchMock.post("*", { body, headers });
        const promise = requestManager.fetch(requestData);
        jest.runOnlyPendingTimers();
        response = await promise;
      });

      afterAll(() => {
        fetchMock.restore();
      });

      it("then the request manager should send the correct request", () => {
        expect(fetchMock.lastCall()).toMatchSnapshot();
      });

      it("then the request manager should return the correct response data", () => {
        expect(response).toMatchSnapshot();
      });
    });

    describe("when multiple requests are made within a batch interval", () => {
      let response: MaybeRawResponseData[];

      beforeAll(async () => {
        jest.useFakeTimers();

        requestManager = await RequestManager.init({
          batch: true,
          fetchTimeout: 10000,
          url: URL,
        });

        const initialRequestData = getRequestData(githubParsedQueries.singleTypeWithFilter.initial);
        const updatedRequestData = getRequestData(githubParsedQueries.singleTypeWithFilter.updated);

        const body = {
          batch: {
            [initialRequestData.hash]: { data: githubQueryResponses.singleTypePartialAndFilter.initial.data },
            [updatedRequestData.hash]: { data: githubQueryResponses.singleTypePartialAndFilter.updated.data },
          },
        };

        const headers = { "cache-control": "public, max-age=5" };
        fetchMock.post("*", { body, headers });

        const promises = [
          requestManager.fetch(initialRequestData),
          requestManager.fetch(updatedRequestData),
        ];

        jest.runOnlyPendingTimers();
        response = await Promise.all(promises);
      });

      afterAll(() => {
        fetchMock.restore();
      });

      it("then the request manager should send the correct request", () => {
        expect(fetchMock.lastCall()).toMatchSnapshot();
      });

      it("then the request manager should return the correct response data", () => {
        expect(response[0]).toMatchSnapshot();
        expect(response[1]).toMatchSnapshot();
      });
    });
  });
});
