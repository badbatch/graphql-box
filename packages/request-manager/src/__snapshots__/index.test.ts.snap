// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@handl/request-manager when batching is set to false then the request manager should return the correct response data 1`] = `
Object {
  "data": Object {
    "organization": Object {
      "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
      "email": "",
      "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      "login": "facebook",
      "name": "Facebook",
      "url": "https://github.com/facebook",
    },
  },
  "headers": Headers {
    "_headers": Object {
      "cache-control": Array [
        "public, max-age=5",
      ],
      "content-length": Array [
        "268",
      ],
      "content-type": Array [
        "application/json",
      ],
    },
  },
}
`;

exports[`@handl/request-manager when batching is set to false then the request manager should send the correct request 1`] = `
Array [
  "https://api.github.com/graphql?requestId=651f14b3df12bef942ed3b1ba9e7f1f1",
  Object {
    "body": "{\\"batched\\":false,\\"request\\":\\"\\\\n  query {\\\\n    organization(login: \\\\\\"facebook\\\\\\") {\\\\n      description\\\\n      email\\\\n      login\\\\n      name\\\\n      url\\\\n      id\\\\n    }\\\\n  }\\\\n\\"}",
    "headers": Headers {
      "_headers": Object {
        "content-type": Array [
          "application/json",
        ],
      },
    },
    "method": "POST",
  },
]
`;

exports[`@handl/request-manager when batching is set to true when a single request is made within a batch interval then the request manager should return the correct response data 1`] = `
Object {
  "data": Object {
    "organization": Object {
      "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
      "email": "",
      "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      "login": "facebook",
      "name": "Facebook",
      "url": "https://github.com/facebook",
    },
  },
  "headers": Headers {
    "_headers": Object {
      "cache-control": Array [
        "public, max-age=5",
      ],
      "content-length": Array [
        "315",
      ],
      "content-type": Array [
        "application/json",
      ],
    },
  },
}
`;

exports[`@handl/request-manager when batching is set to true when a single request is made within a batch interval then the request manager should send the correct request 1`] = `
Array [
  "https://api.github.com/graphql?requestId=651f14b3df12bef942ed3b1ba9e7f1f1",
  Object {
    "body": "{\\"batched\\":true,\\"request\\":{\\"651f14b3df12bef942ed3b1ba9e7f1f1\\":\\"\\\\n  query {\\\\n    organization(login: \\\\\\"facebook\\\\\\") {\\\\n      description\\\\n      email\\\\n      login\\\\n      name\\\\n      url\\\\n      id\\\\n    }\\\\n  }\\\\n\\"}}",
    "headers": Headers {
      "_headers": Object {
        "content-type": Array [
          "application/json",
        ],
      },
    },
    "method": "POST",
  },
]
`;

exports[`@handl/request-manager when batching is set to true when multiple requests are made within a batch interval then the request manager should return the correct response data 1`] = `
Object {
  "data": Object {
    "organization": Object {
      "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
      "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      "login": "facebook",
    },
  },
  "headers": Headers {
    "_headers": Object {
      "cache-control": Array [
        "public, max-age=5",
      ],
      "content-length": Array [
        "414",
      ],
      "content-type": Array [
        "application/json",
      ],
    },
  },
}
`;

exports[`@handl/request-manager when batching is set to true when multiple requests are made within a batch interval then the request manager should return the correct response data 2`] = `
Object {
  "data": Object {
    "organization": Object {
      "email": "",
      "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      "name": "Facebook",
      "url": "https://github.com/facebook",
    },
  },
  "headers": Headers {
    "_headers": Object {
      "cache-control": Array [
        "public, max-age=5",
      ],
      "content-length": Array [
        "414",
      ],
      "content-type": Array [
        "application/json",
      ],
    },
  },
}
`;

exports[`@handl/request-manager when batching is set to true when multiple requests are made within a batch interval then the request manager should send the correct request 1`] = `
Array [
  "https://api.github.com/graphql?requestId=8fc321b1bf4d18ee3d1f2d745b4b28f8-6f657b2d4a58985d1a26dbf3176137b0",
  Object {
    "body": "{\\"batched\\":true,\\"request\\":{\\"8fc321b1bf4d18ee3d1f2d745b4b28f8\\":\\"\\\\n    query {\\\\n      organization(login: \\\\\\"facebook\\\\\\") {\\\\n        description\\\\n        login\\\\n        id\\\\n      }\\\\n    }\\\\n  \\",\\"6f657b2d4a58985d1a26dbf3176137b0\\":\\"\\\\n    query {\\\\n      organization(login: \\\\\\"facebook\\\\\\") {\\\\n        email\\\\n        name\\\\n        url\\\\n        id\\\\n      }\\\\n    }\\\\n  \\"}}",
    "headers": Headers {
      "_headers": Object {
        "content-type": Array [
          "application/json",
        ],
      },
    },
    "method": "POST",
  },
]
`;
